### PR TITLE
Add GCC 9.4, 9.5, 10.4 for Arm and AArch64 and And Morello Alpha 2

### PR DESCRIPTION
**Pull request by TamarChristinaArm**
Final CC part for new toolchains

Related infra patch https://github.com/compiler-explorer/infra/pull/935
Related cross-gcc patch https://github.com/compiler-explorer/gcc-cross-builder/pull/50

### CODE REVIEW

**Review by mattgodbolt — COMMENTED** [review]
So the arm 64 ones are OK but somehow the ARM ones aren't installing:

```
$ sudo ce_install install compilers/c++/cross/gcc/arm 9.4.0
2023-01-27 14:06:42,608 lib.installation_context INFO     Making uncached requests
2023-01-27 14:06:42,608 lib.ce_install  INFO     Creating multiprocessing pool with 2 workers
2023-01-27 14:06:44,092 compilers/dotnet trunk INFO     Most recent dotnet-trunk is 20230111
Installing compilers/c++/cross/gcc/arm 9.4.0
2023-01-27 14:06:47,381 lib.installation_context INFO     Fetching https://s3.amazonaws.com/compiler-explorer/opt/arm-gcc-9.4.0.tar.xz (54974768 bytes)
2023-01-27 14:06:49,280 lib.installation_context INFO     100% of https://s3.amazonaws.com/compiler-explorer/opt/arm-gcc-9.4.0.tar.xz
2023-01-27 14:06:49,281 lib.installation_context INFO     Piping to tar Jxf -
2023-01-27 14:08:24,140 lib.installation_context INFO     Moving from staging (/opt/compiler-explorer/staging/9861153a-bc28-488f-be07-00eceb227861/gcc-9.4.0) to final destination (/opt/compiler-explorer/arm/gcc-9.4.0)
2023-01-27 14:08:37,304 lib.installation_context INFO     Destination /opt/compiler-explorer/arm/gcc-9.4.0 exists, temporarily moving out of the way (to /opt/compiler-explorer/staging/9861153a-bc28-488f-be07-00eceb227861/temp_orig)
2023-01-27 14:08:58,162 lib.ce_install  ERROR    compilers/c++/cross/gcc/arm 9.4.0 installed OK, but doesn't appear as installed after
0 packages installed OK, 0 skipped, and 1 failed installation
Failed:
  compilers/c++/cross/gcc/arm 9.4.0
```

It looks like maybe the extraction path has changed?

```
$ ls /opt/compiler-explorer/arm/gcc-9.4.0
aarch64-unknown-linux-gnu
 ls /opt/compiler-explorer/arm/gcc-9.4.0/aarch64-unknown-linux-gnu/aarch64-unknown-linux-gnu/
bin  debug-root  include  lib  lib64  sysroot
$ ls /opt/compiler-explorer/arm/gcc-9.4.0/aarch64-unknown-linux-g
nu/aarch64-unknown-linux-gnu/bin
ar  as  ld  ld.bfd  nm  objcopy  objdump  ranlib  readelf  strip
$ find /opt/compiler-explorer/arm/gcc-9.4.0/aarch64-unknown-linux-gnu/ -name \*g++
/opt/compiler-explorer/arm/gcc-9.4.0/aarch64-unknown-linux-gnu/bin/aarch64-unknown-linux-gnu-g++
```

so - we would need to update the "test executable" in in the infra, and then it'll install OK, but we'll need to change the paths here (unless it's an unexpected difference you'd like to investigate).

Let me know! :)

**Comment by TamarChristinaArm** [comment]
> $ ls /opt/compiler-explorer/arm/gcc-9.4.0
> aarch64-unknown-linux-gnu

Hmm this is confusing, this looks like it build an AArch64 compiler instead of an Arm one.
The config file definitely says Arm, https://github.com/compiler-explorer/gcc-cross-builder/blob/main/build/latest/arm-9.4.0.config#L142 but it has a weird entry setting 64-bit to True as well.. https://github.com/compiler-explorer/gcc-cross-builder/blob/main/build/latest/arm-9.4.0.config#L176

> (unless it's an unexpected difference you'd like to investigate).

It is, I will investigate, sorry for the mixup! looks like something went wrong during upgrading of the config to new crossng.

**Comment by mattgodbolt** [comment]
No worries at all Tamar! This is why we have all the checks in place. Let me know as and when I can rebuild things or if you need me to poke about on my side at all!

**Comment by dkm** [comment]
shouldn't have aarch64 in the triple for the arm 32bits compiler. Looks like CT_ARCH_32 should be set here, not CT_ARCH_64 https://github.com/compiler-explorer/gcc-cross-builder/blob/5172560b13ffadcbf28a11a40b5ccc0403c659e9/build/latest/arm-9.4.0.config#L176 ? At least that's not the case for other compiler.

**Comment by dkm** [comment]
Woops, missed the replies while writing mine, sorry for the dup

**Comment by Mistuke** [comment]
> Woops, missed the replies while writing mine, sorry for the dup

No problem at all, looks like `/opt/crosstool-ng-latest/bin/ct-ng upgradeconfig` did some unexpected things. I'll inspect the changes.

**Comment by TamarChristinaArm** [comment]
> Woops, missed the replies while writing mine, sorry for the dup

No problem at all, looks like /opt/crosstool-ng-latest/bin/ct-ng upgradeconfig did some unexpected things. I'll inspect the changes.

**Comment by TamarChristinaArm** [comment]
Sorry for the long delay. I've been looking at this today, I think it's only the 9.4.0 Arm one that's broken. Is that right? I don't see any failing checks in the above

**Comment by dkm** [comment]
> Sorry for the long delay. I've been looking at this today, I think it's only the 9.4.0 Arm one that's broken. Is that right? I don't see any failing checks in the above

True, I've checked paths of installed compilers, and currently the following are not installed:
 - ` /opt/compiler-explorer/arm/gcc-9.4.0/arm-unknown-linux-gnueabihf/bin/arm-unknown-linux-gnueabihf-g++`
 - ` /opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp2-x86_64/bin/aarch64-none-elf-gcc`
Looking now why morello isn't installed.

**Comment by dkm** [comment]
There's an extra sub dir:
```
$ tree -L 1 /opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp1-x86_64/
/opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp1-x86_64/
├── aarch64-none-elf
├── bin
├── include
├── lib
├── libexec
└── share

6 directories, 0 files

 $ tree -L 2 /opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp2-x86_64/
/opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp2-x86_64/
└── arm-gnu-toolchain-10.1.morello-alp2-x86_64-aarch64-none-elf
    ├── 10.1.morello-alp2-x86_64-aarch64-none-elf-manifest.txt
    ├── aarch64-none-elf
    ├── bin
    ├── include
    ├── lib
    ├── libexec
    └── share
```

@TamarChristinaArm Is it expected?

**Comment by TamarChristinaArm** [comment]
> There's an extra sub dir:
> 
> @TamarChristinaArm Is it expected?

Arggg no it's not... I've asked the team that makes the package to watch the format.  For this one I'll have to fix it.. Is it better to get rid of the indirection on the Infra side?

**Comment by dkm** [comment]
> > There's an extra sub dir:
> > @TamarChristinaArm Is it expected?
> 
> Arggg no it's not... I've asked the team that makes the package to watch the format. For this one I'll have to fix it.. Is it better to get rid of the indirection on the Infra side?

I guess this would be the best option, but I don't know how to fix that. There may be some dark yaml magic incantation for this, but I don't know it (yet). @mattgodbolt is probably our only hope here :) Looking at the untar python code, it doesn't look like there's a way to strip an arbitrary path component.

**Comment by TamarChristinaArm** [comment]
Hmm yeah if there's no dark magic incantations to strip the path, I can fix it here then. But I'll wait what Matt says, I'm hoping to keep things consistent here.

**Comment by TamarChristinaArm** [comment]
In the meantime, https://github.com/compiler-explorer/gcc-cross-builder/pull/51 fixes the Arm 9.4.0 release.

**Comment by dkm** [comment]
Oh, right :) Reopened !

**Comment by dkm** [comment]
FYI, 9.4.0 has been installed correctly after a rebuild using your fix.

**Comment by TamarChristinaArm** [comment]
Yay, one down, one to go :)

**Comment by mattgodbolt** [comment]
> I guess this would be the best option, but I don't know how to fix that. There may be some dark yaml magic incantation for this, but I don't know it (yet). @mattgodbolt is probably our only hope here :) Looking at the untar python code, it doesn't look like there's a way to strip an arbitrary path component.

Hi friends...good question. I could have sworn there's a way to do this but I'm having trouble seeing it. There may be a combination of `subdir`, `untar_dir` and `path_name` that can do it but it's ugly code. I'll have to take a look later. Sorry it's not clear cut (no `--strip-components` support... argh)

**Comment by dkm** [comment]
I'll have a look to add support for `strip_component` in yaml (if it's not already here but well hidden)

**Comment by dkm** [comment]
Finally took the time (sorry, forgot about this). We do have `strip_components` support... so that was an easy fix. I'll wait for basic tests to finish then I'll merge (in infra)

**Comment by dkm** [comment]
Merged in infra. I'll check that both compilers are correctly installed tomorrow and rereview/merge this PR if OK

**Comment by TamarChristinaArm** [comment]
@dkm awesome! many thanks!

**Comment by dkm** [comment]
```
$ tree -L 1 /opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp2-x86_64/
/opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp2-x86_64/
├── 10.1.morello-alp2-x86_64-aarch64-none-elf-manifest.txt
├── aarch64-none-elf
├── bin
├── include
├── lib
├── libexec
└── share
 $ /opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp2-x86_64/bin/aarch64-none-elf-gcc -v
Using built-in specs.
COLLECT_GCC=/opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp2-x86_64/bin/aarch64-none-elf-gcc
....
$  /opt/compiler-explorer/arm64/gcc-9.4.0/aarch64-unknown-linux-gnu/bin/aarch64-unknown-linux-gnu-gcc -v
Using built-in specs.
COLLECT_GCC=/opt/compiler-explorer/arm64/gcc-9.4.0/aarch64-unknown-linux-gnu/bin/aarch64-unknown-linux-gnu-gcc
....
$ /opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp1-x86_64/bin/aarch64-none-elf-gcc -v
Using built-in specs.
COLLECT_GCC=/opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp1-x86_64/bin/aarch64-none-elf-gcc
...
```

Looks correct now. Will read again (later) the initial messages above before finally merging this, checking I didn't miss something  :)

**Comment by TamarChristinaArm** [comment]
how'd it go @dkm ? :)

**Comment by dkm** [comment]
(on my todo list, sorry for the delay)

File: etc/config/c++.amazon.properties
Comment (dkm) [inline etc/config/c++.amazon.properties:1328]: Note: exe not present.

Comment (dkm) [reply etc/config/c++.amazon.properties:1328]: Probably should be: `/opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp2-x86_64/bin/aarch64-none-elf-c++filt`

Comment (dkm) [reply etc/config/c++.amazon.properties:1328]:
```suggestion
group.gcc64marm.objdumper=/opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp2-x86_64/bin/aarch64-none-elf-objdump
group.gcc64marm.demangler=/opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp2-x86_64/bin/aarch64-none-elf-c++filt
```

File: etc/config/c++.amazon.properties
Comment (dkm) [inline etc/config/c++.amazon.properties:1327]:
```suggestion
group.gcc64marm.objdumper=/opt/compiler-explorer/arm64morello/gcc-aarch64-none-elf-10.1.morello-alp2-x86_64/bin/aarch64-none-elf-objdump
```

**Review by dkm — CHANGES_REQUESTED** [review]
minor change needed, but should be good to go. I may apply them and commit if I have more time later

**Comment by mattgodbolt** [comment]
Thanks @dkm !

**Comment by TamarChristinaArm** [comment]
Awesome, thanks @dkm !

**Comment by jrtc27** [comment]
I'm going to repeat my comments on https://github.com/compiler-explorer/compiler-explorer/pull/3794:

> Curious as to why you've submitted this to compiler-explorer upstream rather than the CHERI instance at cheri-compiler-explorer.cl.cam.ac.uk. Having the split is confusing and unhelpful.

> And given this now means compiler-explorer upstream can produce assembly using the .chericap directive (https://godbolt.org/z/9zvdd1Ecx), does this mean I can upstream my patch to not filter this out when filtering out non-data directives (https://github.com/CTSRD-CHERI/cheri-docker-images/blob/master/cheri-compiler-explorer/patches/chericap-directive.diff)?..

And Alex's too:

> Nice, I was not aware of this. Maybe it would also make sense to add the more mature LLVM-based compilers? Those are currently available on https://cheri-compiler-explorer.cl.cam.ac.uk/

**Comment by dkm** [comment]
Yes you could upstream you filter patch @jrtc27 ! As for the cheri llvm compiler, maybe create a dedicated issue... We already have a long backlog, so having requests in comments is a good way to lose them unfortunately...

**Pull request merged**
